### PR TITLE
Account for global stores in hoisting properly

### PR DESCRIPTION
**Pull request by SingleAccretion**
We were not doing so in all cases, which could lead to reordering a throw with a store.

Small [diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1684999&view=ms.vss-build-web.run-extensions-tab) where we no longer hoist across a global store.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
Assuming CI failures are not related:
1) Linux ARM - #65817.
2) OSX x64 - #67106.
3) Windows ARM64 - same failure mode as [here](https://dev.azure.com/dnceng/public/_build/results?buildId=1684635&view=logs&jobId=bb94f37e-a09b-5b51-09b2-ab78e7aa5103) (which was for a comment-only change).

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib

**Review by EgorBo — APPROVED** [review]

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -6910,23 +6910,27 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
...
                 {
-                    // If the LHS of the assignment has a global reference, then assume it's a global side effect.
-                    GenTree* lhs = tree->AsOp()->gtOp1;
-                    if (lhs->gtFlags & GTF_GLOB_REF)
+                    // Assume all stores except "ASG(non-addr-exposed LCL, ...)" are globally visible.
+                    GenTreeLclVarCommon* lclNode;
+                    bool                 isGloballyVisibleStore;
+                    if (tree->OperIs(GT_ASG) && tree->DefinesLocal(m_compiler, &lclNode))
                     {
+                        isGloballyVisibleStore = m_compiler->lvaGetDesc(lclNode)->IsAddressExposed();
```
[the remark this answers is not in the document]
Comment (SingleAccretion) [reply src/coreclr/jit/optimizer.cpp:6920]: You mean doing `Interlocked.Exchange(ref lcl, 0)`? Yes, `lcl` will be address-exposed in that case.

> perhaps makes sense to keep an assert?

Not sure which one... ?

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/jit/optimizer.cpp:6920]:
```
if (!isGloballyVisibleStore)
    assert(!rhs->OperIs(GT_XADD, GT_XORR, GT_XAND, GT_XCHG, GT_LOCKADD, GT_CMPXCHG, GT_MEMORYBARRIER))
```

[the remark this answers is not in the document]
Comment (SingleAccretion) [reply src/coreclr/jit/optimizer.cpp:6920]: I don't think there is a need. Atomics can never be on the RHS (they're never "location" nodes).

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**